### PR TITLE
Set .spec.schedulerName to ds-controller.

### LINKED ISSUE #64857: `schedulerName` of DaemonSet pods should not be default-scheduler
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind feature

**What happened**:

Create a DaemonSet. It'll create pods with schedulerName=default-scheduler. However, the DaemonSet pods aren't actually scheduled by default-scheduler but DaemonSet controller. 

@k82cn

### PR DESCRIPTION
**Pull request by k82cn**
Signed-off-by: Da K. Ma <klaus1982.cn@gmail.com>

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #64857

**Release note**:
```release-note
Set DaemonSet Pod's `.spec.schedulerName` to "kubernetes.io/daemonset-controller".
```

### CODE REVIEW

File: pkg/controller/daemon/util/daemonset_util.go [hunk 1]
```diff
@@ -99,6 +99,11 @@ func CreatePodTemplate(template v1.PodTemplateSpec, generation *int64, hash stri
 		})
 	}
 
+	if !utilfeature.DefaultFeatureGate.Enabled(features.ScheduleDaemonSetPods) {
+		// If scheduled by daemonset controler, set `.spec.schedulerName` to it.
+		newTemplate.Spec.SchedulerName = "daemonset-controller"
```
Comment (liggitt) [inline pkg/controller/daemon/util/daemonset_util.go]: shouldn't this go in the block where we set the node name if features.ScheduleDaemonSetPods is disabled?

are scheduler names namespaced (prefixed with `kubernetes.io/...`)? `default-scheduler` is a reserved value since that existed when the field was introduced, but it doesn't seem like we should suddenly claim another unnamespaced string

File: pkg/apis/core/v1/defaults.go [hunk 2]
```diff
@@ -178,7 +178,9 @@ func SetDefaults_PodSpec(obj *v1.PodSpec) {
 		period := int64(v1.DefaultTerminationGracePeriodSeconds)
 		obj.TerminationGracePeriodSeconds = &period
 	}
-	if obj.SchedulerName == "" {
+
+	// If nodeName is empty, set schedulerName to default scheduler.
+	if obj.SchedulerName == "" && obj.NodeName == "" {
```
Comment (liggitt) [inline pkg/apis/core/v1/defaults.go]: why is this necessary? doesn't the default scheduler also handle preemption of pods assigned to nodes? is that limited to pods that claim the default scheduler as their scheduler name?

Comment (k82cn) [reply pkg/apis/core/v1/defaults.go]: If `.spec.NodeName` is not empty, we did not know whether this Pod is scheduled by default scheduler. The scheduler will monitor all `not terminated` and `not assigned` Pod to calculate available resource on the node.

File: pkg/controller/daemon/util/daemonset_util.go (hunk 1 above)
Comment (k82cn) [reply pkg/controller/daemon/util/daemonset_util.go]: > shouldn't this go in the block where we set the node name if features.ScheduleDaemonSetPods is disabled?

I'm thinking `.spec.SchedulerName` should be part of template, so I put it here :)

> are scheduler names namespaced (prefixed with kubernetes.io/...)? default-scheduler is a reserved value since that existed when the field was introduced, but it doesn't seem like we should suddenly claim another unnamespaced string

en... I think we should make it namespaced :)

Comment (liggitt) [reply pkg/controller/daemon/util/daemonset_util.go]: @msau42 does setting this to a value other than default-scheduler disrupt anything volume-related (attaching, binding, local volume stuff, etc)?

Comment (msau42) [reply pkg/controller/daemon/util/daemonset_util.go]: The volume components do not look at the pod scheduler name.  

However, if the default scheduler does not process the pod (which is the current behavior when the feature is off), then none of the volume predicates will run (ie volume count, volume zone, volume binding).  And I think that is fine for Daemonsets because you can't really use those volume types in a Daemonset:

Volume count, zone: used by rwo volumes which doesn't work in a Daemonset because they can't be attached to more than one node at a time
Volume binding: used by topology aware PVCs which are currently local volumes and rwo zonal volumes.  Which don't work in a Daemonset because they all share the same PVC and are rwo.

**Comment by k82cn** [comment]
ping :)

File: pkg/controller/daemon/util/daemonset_util.go [hunk 3]
```diff
@@ -99,6 +99,11 @@ func CreatePodTemplate(template v1.PodTemplateSpec, generation *int64, hash stri
 		})
 	}
 
+	if !utilfeature.DefaultFeatureGate.Enabled(features.ScheduleDaemonSetPods) {
```
Comment (liggitt) [inline pkg/controller/daemon/util/daemonset_util.go]: I think it makes more sense to do this here, where we already have the logic to delegate to the default scheduler or not:

https://github.com/kubernetes/kubernetes/blob/16c5ba4114bf6443c2ce8c0104ad520dbcece6ce/pkg/controller/daemon/daemon_controller.go#L993-L1007

File: pkg/apis/core/v1/defaults.go (hunk 2 above)
Comment (liggitt) [reply pkg/apis/core/v1/defaults.go]: please revert this. changing defaults like this can break a HA version-skewed cluster:

1. client creates a pod with a nodeName set via a new API server. API server would persist to etcd without defaulting the scheduler name.
2. client retrieves the pod via an old API server, which returns it with the scheduler name defaulted.
3. client changes something unrelated and submits an update via the new API server. the new API server would reject the request because the incoming pod tried to mutate something in the pod spec (the scheduler name, defaulted by the old API server)

File: pkg/controller/daemon/util/daemonset_util.go (hunk 3 above)
Comment (k82cn) [reply pkg/controller/daemon/util/daemonset_util.go]: done

File: pkg/apis/core/v1/defaults.go (hunk 2 above)
Comment (k82cn) [reply pkg/apis/core/v1/defaults.go]: Got that! done :)

File: pkg/controller/daemon/daemon_controller.go [hunk 4]
```diff
@@ -1002,6 +1002,9 @@ func (dsc *DaemonSetsController) syncNodes(ds *apps.DaemonSet, podsToDelete, nod
 					err = dsc.podControl.CreatePodsWithControllerRef(ds.Namespace, podTemplate,
 						ds, metav1.NewControllerRef(ds, controllerKind))
 				} else {
+					// If pod is scheduled by DaemonSetController, set its '.spec.scheduleName'.
+					podTemplate.Spec.SchedulerName = "kubernetes.io/daemonset-controller"
```
Comment (liggitt) [inline pkg/controller/daemon/daemon_controller.go]: need to copy the podTemplate if you're going to change it, just like in the above branch

Comment (k82cn) [reply pkg/controller/daemon/daemon_controller.go]: done

**Comment by liggitt** [comment]
lgtm, will let sig-scheduling tag

**Comment by k82cn** [comment]
@janetkuo , PTAL :)

**Comment by k82cn** [comment]
ping :)

**Comment by k82cn** [comment]
@bsalamat , @Huang-Wei , is this PR ok for you?

**Review by Huang-Wei — COMMENTED** [review]
/lgtm
